Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs

```diff
@@ -41,15 +41,50 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
 
         if (GetBallisticShots(component) >= component.Capacity) return;
 
-        component.Entities.Add(args.Used);
-        component.Container.Insert(args.Used);
+        if (TryComp<BallisticAmmoProviderComponent>(args.Used, out var ammoComp))
+        {
+            if (GetBallisticShots(ammoComp) == 0) return;
+
+            while (GetBallisticShots(component) < component.Capacity && GetBallisticShots(ammoComp) > 0)
```

Comment:
Isn't this going to continue to iterate a container over and over? Seems much better to track it yourself.